[Feature] Introduce Column Hash Predicate as a Record Predicate for data filtering

srlch opened the pull request:
## What I'm doing:
This is a prerequisite for tablet splitting. We introduce `RecordPredicate` and its derived class `ColumnHashIsCongruent` into `RowsetMetadataPB` to provide ability to filter data in all segments of the rowset using this predicate. `RecordPredicate` can be considered as a storage layer only predicate to filter data and `ColumnHashIsCongruent` calculate the available row for a given chunk in the following way:
1. Find columns by hash_column_names in chunk and calculate their combined `crc32` value.
2. `select[i] = (hashes[i] % modulus == remainder)`

Fixes #59134

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

xiangguangyxg commented:
does it support using a hidden generated column to save the hash value ?

srlch commented:
> does it support using a hidden generated column to save the hash value ?

I think this predicate can support it as long as we give the correct col_name of the generated column, we can skip the hash value calculation and just using the column itself.

alvin-celerdata commented on gensrc/proto/types.proto:
```diff
@@ -181,6 +181,18 @@ message IsNullPredicatePB {
     optional bool is_not_null = 2;
 }
 
+message ColumnHashModulusPredicatePB {
+    enum ColumnHashTypePB {
+        UNKNOWN = 0;
+        CRC32 = 1;
+    }
+
+    optional ColumnHashTypePB type = 1;
+    optional int64 modulus = 2;
+    optional int64 remainder = 3;
+    repeated string hash_column_names = 4;
+}
```
This class is too specific.
If a rowset can handle multiple remainders, how can you achieve it?

srlch replied:
How about define a wrapper class out of this class if we need multiple remainders or even modulus.

srlch replied:
I think another option maybe introduce repeated `<modulus, remainder>` pair and `OpType` for `AND/OR`

srlch replied:
Introduce tree structure for such predicate, i think keep very specified predicate simpler is a better choice? We can build a tree to achieve this goal.

alvin-phoenix-ai reviewed (dismissed):
After the definition has been agreed, I will dismiss the request-change.

alvin-celerdata commented on gensrc/proto/lake_types.proto:
```diff
@@ -110,6 +110,7 @@ message RowsetMetadataPB {
     // only be set > 0 if `overlapped` is true
     optional uint32 next_compaction_offset = 13;
     repeated int64 bundle_file_offsets = 14;
+    optional ColumnHashModulusPredicatePB column_hash_modulus_predicate = 15;
```
> I'm not pretty sure what general means here. You mean i should introduce general select predicate here instead of a specified one for hash modulus?

Yes. A base class should be put here.

srlch replied:
Updated, introduce `RecordPredicatePB` as a base class which is a kinds of predicate only for storage layer. `SELECT/FILTER` means it expect record should be selected or filtered by this predicate. Besides, i introduce `RecordPredicateTreeNodePB` structure to represent any kinds of combination of `RecordPredicatePB`.

alvin-phoenix-ai reviewed (commented):
You can reference how TExpr is defined.

srlch commented on gensrc/proto/types.proto:
```diff
@@ -181,6 +181,40 @@ message IsNullPredicatePB {
+        AND = 0;
+        OR = 1;
+        LEAF 2;
+    }
```
Besides, i remove the SELECT/FILTER type and introduce NOT nodeType to represent it

srlch replied:
Any suggestion for this updated?

alvin-celerdata commented on gensrc/proto/types.proto:
```diff
@@ -181,6 +181,31 @@ message IsNullPredicatePB {
+        HASH_MODULUS = 1;
+    }
+
+    message ColumnHashModulusPredicatePB {
```
![image](https://github.com/user-attachments/assets/0af3219d-d4b3-4b2a-b2d2-8e9f305c8896)

![image](https://github.com/user-attachments/assets/b21fca95-447f-42e0-87c3-eadab29bfec4)
`ColumnisCongruent` may be a better name

srlch replied:
`ColumnHashIsCongruentPredicatePB`? Seems should emphasize that the dividend is the hash of columns.

xiangguangyxg commented on gensrc/proto/types.proto:
```diff
@@ -181,6 +181,31 @@ message IsNullPredicatePB {
+    }
+
+    optional RecordPredicateTypePB type = 1;
+    // set by some type of predicate (RecordPredicateTypePB is AND/OR/NOT...) which will be implemented in the future
+    repeated RecordPredicatePB children = 2;
```
not used now, could be added in future when needed.

srlch replied:
Fixed

srlch commented on be/src/storage/record_predicate/record_predicate.h:
```diff
@@ -0,0 +1,67 @@
+    bool _inited;
+};
+
+using RecordPredicateSPtr = std::shared_ptr<RecordPredicate>;
```
I think shared_ptr will be more easier? Because this pointer will be set in `SegmentOption` which is a copyable class. But if this pointer is unique_ptr, `SegmentOption` will be  un-copyable and we need to handle this case.

alvin-celerdata replied:
In SegmentOption it can be a shared-ptr, but children should be unique_ptr

srlch replied:
Fixed for the children definition

srlch commented on be/src/storage/record_predicate/record_predicate.h:
```diff
@@ -0,0 +1,67 @@
+    virtual Status get_column_ids(const TabletSchemaCSPtr& tablet_schema, std::set<ColumnId>* column_ids) const {
+        return Status::OK();
+    }
+    virtual Status get_column_ids(const Schema& schema, std::set<ColumnId>* column_ids) const { return Status::OK(); }
```
Removed from base class. But i add a general interface in Helper class to these three function which make the caller more easier to handle.

srlch commented on be/src/storage/record_predicate/record_predicate.h:
```diff
@@ -0,0 +1,67 @@
+*/
+class RecordPredicate {
+public:
+    using RecordPredicateSPtr = std::shared_ptr<RecordPredicate>;
```
Fixed

srlch commented on be/src/storage/record_predicate/record_predicate.h:
```diff
@@ -0,0 +1,66 @@
+
+class RecordPredicateHelper {
+public:
+    static StatusOr<RecordPredicateSPtr> create(const RecordPredicatePB& record_predicate_pb);
```
Updated

alvin-celerdata commented on be/src/storage/record_predicate/record_predicate.h:
```diff
@@ -0,0 +1,67 @@
+    std::vector<RecordPredicateUPtr> _children;
+};
+
+class RecordPredicateHelper {
```
Could you move it to another file? The base class won't change too much, but the helper will change a lot.

srlch replied:
Fixed

srlch commented on be/src/storage/record_predicate/column_hash_is_congruent.cpp:
```diff
@@ -0,0 +1,80 @@
+}
+
+Status ColumnHashIsCongruent::init(const RecordPredicatePB& record_predicate_pb) {
+    const ColumnHashIsCongruentMeta& column_hash_is_congruent_pb = record_predicate_pb.column_hash_is_congruent();
```
Fixed

srlch commented on be/src/storage/record_predicate/column_hash_is_congruent.h:
```diff
@@ -0,0 +1,52 @@
+
+    virtual Status init(const RecordPredicatePB& record_predicate_pb) override;
+
+    virtual std::vector<std::string> column_names() const override { return _column_names; }
```
Fixed

srlch commented on be/test/storage/record_predicate/record_predicate_test.cpp:
Fixed

srlch commented on be/src/storage/record_predicate/record_predicate.cpp:
```diff
@@ -0,0 +1,101 @@
+        break;
+    }
+    default:
+        return Status::InternalError("Unknown record predicate type: " +
```
Fixed

xiangguangyxg commented on gensrc/proto/types.proto:
```diff
@@ -181,6 +181,31 @@ message IsNullPredicatePB {
+            CRC32 = 1;
+        }
+
+        optional ColumnHashTypePB type = 1;
```
not used ?

srlch replied:
fixed

xiangguangyxg commented on be/src/storage/record_predicate/column_hash_is_congruent.h:
```diff
@@ -0,0 +1,50 @@
+private:
+    Status _check_valid_pb(const ColumnHashIsCongruentMeta& column_hash_is_congruent_meta) const;
+
+    HashType _hash_type;
```
not used ?

srlch replied:
fixed

srlch committed:
[Feature] Introduce Column Hash Predicate as a Record Predicate for rowset data filtering

This is a prerequisite for tablet splitting. We introduce RecordPredicate and its derived class ColumnHashModulusPredicate into RowsetMetadataPB to provide ability to filter data in all segments of the rowset using this predicate.  can be considered as a storage layer only pr
edicate to filter data and ColumnHashModulusPredicate calculate the available row for a given chunk in the following way:
1. Find columns by hash_column_names in chunk and calculate their combined CRC32 value.
2. select[i] = (hashes[i] % modulus == remainder)

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>
